core/drivers/crypto/ele/ele.c

```diff
@@ -395,45 +391,65 @@ TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
 		uint16_t key_size;
 		uint16_t ctx_size;
 		uint32_t crc;
-	} __packed cmd = { };
+	} __packed cmd = {};
 	struct imx_mu_msg msg = {
 		.header.version = ELE_VERSION_BASELINE,
 		.header.size = SIZE_MSG_32(cmd),
 		.header.tag = ELE_REQUEST_TAG,
 		.header.command = ELE_CMD_DERIVE_KEY,
 	};
+	struct imx_ele_buf ele_ctx = {};
+	struct imx_ele_buf ele_key = {};
 
-	if (is_fetched)
-		goto out;
+	assert(ctx && key);
 
-	/*
-	 * Intermediate msb and lsb values are needed. Directly using
-	 * key_addr_msb and key_addr_lsb might be unaligned because of the
-	 * __packed attribute of key_derive_cmd {}
-	 */
-	reg_pair_from_64((uint64_t)virt_to_phys(key), &msb, &lsb);
+	if (key_size != 16 && key_size != 32)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	cmd.key_addr_lsb = lsb;
-	cmd.key_addr_msb = msb;
-	cmd.key_size = HW_UNIQUE_KEY_LENGTH;
+	res = imx_ele_buf_alloc(&ele_ctx, ctx, ctx_size);
+	if (res)
+		goto out;
 
-	reg_pair_from_64((uint64_t)virt_to_phys((void *)pattern), &msb, &lsb);
+	res = imx_ele_buf_alloc(&ele_key, key, key_size);
+	if (res)
+		goto out;
+
+	cmd.key_addr_lsb = ele_key.paddr_lsb;
+	cmd.key_addr_msb = ele_key.paddr_msb;
+	cmd.key_size = key_size;
 
-	cmd.ctx_addr_lsb = lsb;
-	cmd.ctx_addr_msb = msb;
-	cmd.ctx_size = sizeof(pattern);
+	cmd.ctx_addr_lsb = ele_ctx.paddr_lsb;
+	cmd.ctx_addr_msb = ele_ctx.paddr_msb;
+	cmd.ctx_size = ctx_size;
 
 	memcpy(msg.data.u8, &cmd, sizeof(cmd));
 	update_crc(&msg);
 
-	cache_operation(TEE_CACHEFLUSH, key, HW_UNIQUE_KEY_LENGTH);
-	cache_operation(TEE_CACHECLEAN, (void *)pattern, sizeof(pattern));
-
 	res = imx_ele_call(&msg);
 	if (res)
-		panic("failed to get the huk");
+		goto out;
+
+	res = imx_ele_buf_copy(&ele_key, key, key_size);
+out:
+	imx_ele_buf_free(&ele_key);
+	imx_ele_buf_free(&ele_ctx);
+
+	return res;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	static const char pattern[] = "TEE_for_HUK_ELE";
+	static uint8_t key[HW_UNIQUE_KEY_LENGTH];
+	static bool is_fetched;
+
+	if (is_fetched)
+		goto out;
+
+	if (imx_ele_derive_key((const uint8_t *)pattern, sizeof(pattern), key,
```

Comment:
Am I right to assume that the HUK will remain the same, it's only the way you're obtaining it that has changed?

Reply:
Yes, you are right.